### DISCUSSION TITLE
Possible to use shader variant options/macro's to control data definitions in material SRG?

### DISCUSSION THREAD

**Discussion by ohmaya** [o3de/o3de#12168] [Help Wanted]
Looks like compile-time branch like the following can only appear in shader code:
`    if (o_useRed) {
        color = float3(1, 0, 0);
    } else {
        color = float3(0, 0, 1);
    }`
... but not in SRG definitions?
`ShaderResourceGroup MaterialSrg : SRG_PerMaterial
{
    if (o_useRed)
    {
        float3 m_color;
    }`

It is possible to define a macro in .azsl, but using that macro in the material SRG results in an error, then I have to always define that constant in the material SRG.

**Comment by siliconvoodoo** [comment]
What you can probably do is have one .azsl and multiple .shader and put your macro values in the .shader JSON.
It's documented here how to: https://github.com/o3de/o3de/wiki/%5BAtom%5D-Developer-Guide:-Shader-Build-Arguments-Customization

**Comment by galibzon** [comment]
The best solution is to define a shader Supervariant inside the *.shader file to pass custom Macro definitions. The AZSL file would look like this:
```
ShaderResourceGroup MaterialSrg : SRG_PerMaterial { 
#ifdef USE_RED
       float3 m_color;
#endif
 }
```

This would be part of the content of the .shader file, you'll add a new supervariant called `myRedVersion` which will pass the correct macro definition.
```json
    "Supervariants":
    [
        {
            "Name": "myRedVersion",
            "AddBuildArguments": {
                "defines": ["USE_RED"]
            }
        }
    ]
```

**Comment by santorac** [comment]
It depends on what exactly you are trying to do. You should be able to use the preprocessor however you like inside the shader code, to remove certain ShaderResourceGroup fields. But the trick is you have to find an appropriate place to set the option. The posts above describe a couple ways. Here are all the ways you can set compile-time flags for the shader preprocessor:

1. Modify a shader_build_options.json file (this is what @siliconvoodoo was talking about). This is the approach you want to take if you are trying to generally impact the entire shader compilation pipeline, like customizing the behavior for a specific target platform.
2. Use supervariants in the .shader file (this is what @galibzon was talking about). This approach requires the runtime to make some explicit calls in C++ to select an appropriate supervariant. We currently only use this for configuring MSAA vs non-MSAA shaders.
3. Define the flags in the .shader file directly. For example, https://github.com/o3de/o3de/blob/04e4ce3acc00fb859e368f4c15f37234d234a194/Gems/Atom/Feature/Common/Assets/Shaders/Materials/StandardPBR/StandardPBR_ForwardPass.shader#L4
4. Make some think .azsl file that defines some flag and then includes another .azsli file. For example, https://github.com/o3de/o3de/blob/04e4ce3acc00fb859e368f4c15f37234d234a194/Gems/Atom/Feature/Common/Assets/Shaders/Materials/StandardPBR/StandardPBR_DepthPass_WithPS.azsl#L21

That being said, you are specifically asking about MaterialSrg. The main requirement here, and likely the issue you are running into, is that every shader in a material type must use exactly the same MaterialSrg. This is a core design restriction, and is enforced here: https://github.com/o3de/o3de/blob/04e4ce3acc00fb859e368f4c15f37234d234a194/Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialTypeAssetCreator.cpp#L60

If you want to use preprocessor flags to remove parts of the MaterialSrg, you just need to make sure you set the flag to the same value for every shader in the material type.
